clients/drcachesim/simulator/caching_device.h

```diff
@@ -161,6 +204,16 @@ class caching_device_t {
     addr_t last_tag_;
     int last_way_;
     int last_block_idx_;
+    // Optimization: keep a hashtable for quick lookup of {block,way}
+    // given a tag, if using a large cache hierarchy where serial
+    // walks over the associativity end up as bottlenecks.
+    // We can't easily remove the blocks_ array and replace with just
+    // the hashtable as replace_which_way(), etc. want quick access to
+    // every way for a given line index.
+    std::unordered_map<addr_t, std::pair<caching_device_block_t *, int>,
```

Comment:
Two comments:

1. absl containers are open-source, have we considered migrating to them? I wonder how much perf we could get with that. Obviously not for this PR of course.
2. Is it necessary to provide the hash function signature? IIUC `std::hash<addr_t>` is expected to return `std::size_t`.

Reply:
1. That is worth considering.  There are other places where the std tables were sub-optimal: I think it was raw2trace.  Here we go: https://github.com/DynamoRIO/dynamorio/issues/2056#issuecomment-333167409.  I filed #4844 on absl.

2. It seems to be required for the hash function as a lambda.